subscription_manager.go

```diff
@@ -0,0 +1,271 @@
+// manager listens on. It must only be called when the subscription manager is
+// stopped.
+func (sm *subscriptionManager) ResetSubscribeChan(subscribeCh <-chan []jobcompleter.CompleterJobUpdated) {
+	sm.subscribeCh = subscribeCh
+}
+
+func (sm *subscriptionManager) Start(ctx context.Context) error {
+	sm.wg.Add(1)
+	go func() {
+		defer sm.wg.Done()
+
+		for updates := range sm.subscribeCh {
+			sm.distributeJobEvents(updates)
+		}
+	}()
+
+	return nil
+}
+
+func (sm *subscriptionManager) Stop() {
+	sm.wg.Wait()
+
+	// Remove all subscriptions and close corresponding channels.
+	func() {
+		sm.mu.Lock()
+		defer sm.mu.Unlock()
+
+		for subID, sub := range sm.subscriptions {
+			close(sub.Chan)
+			delete(sm.subscriptions, subID)
+		}
+	}()
+}
+
+func (sm *subscriptionManager) logStats(ctx context.Context, svcName string) {
+	// TODO: don't use mutex + struct internals externally like this
+	sm.statsMu.Lock()
+	defer sm.statsMu.Unlock()
+
+	sm.Logger.InfoContext(ctx, svcName+": Job stats (since last stats line)",
+		"num_jobs_run", sm.statsNumJobs,
+		"average_complete_duration", sm.safeDurationAverage(sm.statsAggregate.CompleteDuration, sm.statsNumJobs),
+		"average_queue_wait_duration", sm.safeDurationAverage(sm.statsAggregate.QueueWaitDuration, sm.statsNumJobs),
+		"average_run_duration", sm.safeDurationAverage(sm.statsAggregate.RunDuration, sm.statsNumJobs))
+
+	sm.statsAggregate = jobstats.JobStatistics{}
+	sm.statsNumJobs = 0
+}
+
+// Handles a potential divide by zero.
+func (sm *subscriptionManager) safeDurationAverage(d time.Duration, n int) time.Duration {
+	if n == 0 {
+		return 0
+	}
+	return d / time.Duration(n)
+}
+
+// Receives updates from the completer and prompts the client to update
+// statistics and distribute jobs into any listening subscriber channels.
+// (Subscriber channels are non-blocking so this should be quite fast.)
+func (sm *subscriptionManager) distributeJobEvents(updates []jobcompleter.CompleterJobUpdated) {
+	func() {
+		sm.statsMu.Lock()
+		defer sm.statsMu.Unlock()
+
+		for _, update := range updates {
+			stats := update.JobStats
+			sm.statsAggregate.CompleteDuration += stats.CompleteDuration
+			sm.statsAggregate.QueueWaitDuration += stats.QueueWaitDuration
+			sm.statsAggregate.RunDuration += stats.RunDuration
+			sm.statsNumJobs++
+		}
+	}()
+
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	// Quick path so we don't need to allocate anything if no one is listening.
+	if len(sm.subscriptions) < 1 {
+		return
+	}
+
+	for _, update := range updates {
+		sm._distributeJobEvent(update.Job, jobStatisticsFromInternal(update.JobStats))
+	}
+}
+
+// Distribute a single event into any listening subscriber channels.
+//
+// Job events should specify the job and stats, while queue events should only specify
+// the queue.
+//
+// MUST be called with sm.mu already held.
+func (sm *subscriptionManager) _distributeJobEvent(job *rivertype.JobRow, stats *JobStatistics) {
```

Comment:
Hmm, can we drop this underscore convention? I don't think it's really a thing. Out of 22.5k struct functions in the Go codebase, I did find three instances of one, but they're all kind of weird special cases:

```
$ ag --no-break --nofilename 'func \([^)]+\) [A-Za-z]' | wc -l
   22438

$ ag --no-break --nofilename 'func \([^)]+\) _[A-Za-z]' | wc -l
       3
```

```
src/go/types/scope.go
119:func (s *Scope) _InsertLazy(name string, resolve func() Object) bool {

src/runtime/symtab.go
751:func (f funcInfo) _Func() *Func {

src/vendor/golang.org/x/text/unicode/norm/input.go
32:func (in *input) _byte(p int) byte {
```

How about renaming:

* `distributeJobEvents` -> `distributeJobUpdates`
* `_distributeJobEvent` -> `distributeJobEvent` 

Makes them a little more distinct.

If the concern is other places calling private members of this class, the underscore doesn't stop it, and the right fix is to try and get this into its own package.

Reply:
Can't recall where I picked up that convention (Miz maybe?) but I believe I've seen it used for indicating methods which are meant to be called only when a mutex is already held. No problem with changing to your suggestions though.